src/RetrySocketConnector.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace Amp\Socket;
+
+use Amp\Cancellation;
+use function Amp\delay;
+
+final class RetrySocketConnector implements SocketConnector
+{
+    public function __construct(
+        private readonly SocketConnector $delegate,
+    ) {
+    }
+
+    public function connect(
+        string $uri,
+        ?ConnectContext $context = null,
+        ?Cancellation $cancellation = null
+    ): EncryptableSocket {
+        $attempts = 0;
+        $failures = [];
+        $context ??= new ConnectContext;
```

Comment:
Any reason why we initialize the context here? So we don't create a new one for each attempt?

Reply:
That and the object is used below, so it's a way of grabbing the default value.

Reply:
Should the backoff even be part of `ConnectContext` or rather just a constructor parameter for `RetrySocketConnector`?